[jit] improve error messages when a method or attribute is missing

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#27110 [jit] improve error messages when a method or attribute is missing**

Previously missing methods on some types like tensors would talk about
'builtins' which are only a thing inside of the compiler. Furthermore,
the error would only occur when the builtin was applied and it was discovered
that no builtin existed. This changes the error message so that it
discovers that method on our builtin types does not exist on attribute lookup.

Differential Revision: [D17677616](https://our.internmc.facebook.com/intern/diff/D17677616)

zdevito committed:
[jit] improve error messages when a method or attribute is missing

Previously missing methods on some types like tensors would talk about
'builtins' which are only a thing inside of the compiler. Furthermore,
the error would only occur when the builtin was applied and it was discovered
that no builtin existed. This changes the error message so that it
discovers that method on our builtin types does not exist on attribute lookup.

eellison commented on torch/csrc/jit/script/sugared_value.h:
```diff
@@ -168,6 +168,13 @@ struct TORCH_API BuiltinFunction : public SugaredValue {
       at::ArrayRef<NamedValue> inputs,
       size_t n_binders) override;
+
+  // try to create this builtin but if it doesn't exit or the self argument
```
s/exit/exist

eellison commented on torch/csrc/jit/script/sugared_value.cpp:
```diff
@@ -509,6 +511,31 @@ std::shared_ptr<SugaredValue> NamedTupleConstructor::call(
 
+std::shared_ptr<BuiltinFunction> BuiltinFunction::tryCreate(
+    Symbol symbol,
+    c10::optional<NamedValue> self) {
```
Will `self` ever be `c10::nullopt` here ?

zdevito replied:
Not in the current PR, but it seems like a good idea to have the arguments to tryCreate match the arguments to BuiltinFunction::create so that the description of what it does is clear.

eellison commented on torch/csrc/jit/script/sugared_value.h:
```diff
@@ -168,6 +168,13 @@ struct TORCH_API BuiltinFunction : public SugaredValue {
       size_t n_binders) override;
+
+  // try to create this builtin but if it doesn't exit or the self argument
+  // cannot possibly match, the return nullptr. Use in situations where it is
```
s/the return/return

eellison commented on torch/csrc/jit/script/sugared_value.cpp:
```diff
@@ -509,6 +511,31 @@ std::shared_ptr<SugaredValue> NamedTupleConstructor::call(
+      return std::make_shared<BuiltinFunction>(symbol, nullptr);
+    }
+    if (auto index = op->schema().argumentIndexWithName("self")) {
+      std::unordered_map<std::string, TypePtr> type_env;
```
Could we just reuse `tryMatchArgument` here? That way this would also work with conversions.

zdevito replied:
Since this is for method invocation where the method should be determined by the self argument, we do not want to apply conversions to self. Methods should only exist if the type matches without conversion. Our previous implementation was too accepting and open to weird bugs here. Furthermore, calling tryMatchArgument modifies the graph, which we do not want to do here because the call has not been made yet. It is possible that the call ends up being invoked multiple times, or not at all.

eellison replied:
Not applying conversions makes since here. I don't see why we can't use `tryMatchArgument` with no conversions.

It's already the case that we modify the graph for schemas we don't match to; any extra nodes get DCE'd.

zdevito replied:
tryMatchArgument is private, takes a lot of arguments, and would be awkward to expose. It is not something that should be called outside of schema matching. Even just accessing `self` as a Value in this context requires potentially modifying a graph because self is NamedValue (potentially needs to be inserted as a constant). Refactoring all of tryMatchArgument so that it doesn't modify the graph would be difficult. I don't like the fact that we rely on DCE here and I don't want to propagate it to more places. It makes debugging harder.

eellison replied:
I think it's a little weird to fix inserting unmatched nodes only for this specific case. i don't know how invasive the more general fix will be though.

eellison reviewed (commented):
Looks good, a couple comments

zdevito committed:
Update on "[jit] improve error messages when a method or attribute is missing"

[jit] improve error messages when a method or attribute is missing

Previously missing methods on some types like tensors would talk about
'builtins' which are only a thing inside of the compiler. Furthermore,
the error would only occur when the builtin was applied and it was discovered
that no builtin existed. This changes the error message so that it
discovers that method on our builtin types does not exist on attribute lookup.

gh-metadata: pytorch pytorch 27110 gh/zdevito/120/head

zdevito committed:
Update on "[jit] improve error messages when a method or attribute is missing"

[jit] improve error messages when a method or attribute is missing

Previously missing methods on some types like tensors would talk about
'builtins' which are only a thing inside of the compiler. Furthermore,
the error would only occur when the builtin was applied and it was discovered
that no builtin existed. This changes the error message so that it
discovers that method on our builtin types does not exist on attribute lookup.

gh-metadata: pytorch pytorch 27110 gh/zdevito/120/head
